File: 0082-Remove-Duplicates-from-Sorted-List-II/0082.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        h = ListNode(-1)
        h.next = head
        pre = h
        cur = head
        while cur != None: 
            duplicate = False          
            while cur.next != None and cur.val == cur.next.val:
                cur = cur.next
                duplicate = True
            
            if duplicate == False:
                pre = cur
            else:
                pre.next = cur.next

            cur = cur.next

        return h.next
           
        
def createList():
    head = ListNode(0)
    cur = head
    cur.next = ListNode(0)
    cur.next.next = ListNode(1)
    cur.next.next.next = ListNode(2)
    return head
            
def printList(head):
    cur = head
    while cur != None:
        print(cur.val, '-->', end='')
        cur = cur.next

    print('NULL')

if __name__ == "__main__":
    head = createList()
    printList(head)
    res = Solution().deleteDuplicates(head)
    printList(res)

File: src/0082-Remove-Duplicates-from-Sorted-List-II/0082.cpp
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    ListNode* deleteDuplicates(ListNode* head) 
    {
        ListNode* h = new ListNode(-1);
        h->next = head;
        ListNode* pre = h;
        ListNode* cur = head;
        while (cur != nullptr)
        {
            bool duplicate = false;
            while (cur->next != nullptr && cur->val == cur->next->val)
            {
                ListNode* delNode = cur;
                cur = cur->next;
                delete delNode;
                duplicate = true;
            }
            if (duplicate == false)
            {
                pre = cur;
                cur = cur->next;
            }
            else
            {
                pre->next = cur->next;
                ListNode* delNode = cur;
                cur = cur->next;
                delete delNode;
            }
        }
        ListNode* res = h->next;
        delete h;
        return res;
    }
};

int main()
{
    return 0;
}